pkg/collector/corechecks/containers/containerd.go

```diff
@@ -0,0 +1,359 @@
+
+package containers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/containerd/cgroups"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/typeurl"
+	"github.com/gogo/protobuf/types"
+	"gopkg.in/yaml.v2"
+
+	"github.com/DataDog/datadog-agent/pkg/aggregator"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	core "github.com/DataDog/datadog-agent/pkg/collector/corechecks"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	cutil "github.com/DataDog/datadog-agent/pkg/util/containerd"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	containerdCheckName = "containerd"
+)
+
+// ContainerCheck grabs containerd metrics and events
+type ContainerdCheck struct {
+	core.CheckBase
+	hostname string
+	instance *ContainerdConfig
+	sub      *Subscriber
+}
+
+// ContainerdConfig contains the custom options and configurations set by the user.
+type ContainerdConfig struct {
+	Tags    []string `yaml:"tags"`
+	Filters []string `yaml:"filters"`
+}
+
+func init() {
+	corechecks.RegisterCheck(containerdCheckName, ContainerdFactory)
+}
+
+// ContainerdFactory is used to create register the check and initialize it.
+func ContainerdFactory() check.Check {
+	return &ContainerdCheck{
+		CheckBase: corechecks.NewCheckBase(containerdCheckName),
+		instance:  &ContainerdConfig{},
+		sub:       &Subscriber{},
+	}
+}
+
+// Parse is used to get the configuration set by the user
+func (co *ContainerdConfig) Parse(data []byte) error {
+	if err := yaml.Unmarshal(data, co); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Configure parses the check configuration and init the check
+func (c *ContainerdCheck) Configure(config, initConfig integration.Data) error {
+	var err error
+	if err = c.CommonConfigure(config); err != nil {
+		return err
+	}
+
+	c.hostname, err = util.GetHostname()
+	if err != nil {
+		return err
+	}
+
+	if err = c.instance.Parse(config); err != nil {
+		return err
+	}
+	c.sub.Filters = c.instance.Filters
+	return nil
+}
+
+// Run executes the check
+func (c *ContainerdCheck) Run() error {
+	sender, err := aggregator.GetSender(c.ID())
+	defer sender.Commit()
+	if err != nil {
+		return err
+	}
+	// As we do not rely on a singleton, we ensure connectivity every check run.
+	cu, errHealth := cutil.GetContainerdUtil()
+	if errHealth != nil {
+		log.Infof("Error ensuring connectivity with Containerd daemon %v", errHealth)
+		return errHealth
+	}
+	ns := cu.Namespace()
+
+	if c.sub == nil {
+		c.sub = CreateEventSubscriber("ContainerdCheck", ns, c.instance.Filters)
+	}
+
+	if !c.sub.IsRunning() {
+		// Keep track of the health of the Containerd socket
+		c.sub.CheckEvents(cu)
+	}
+	events := c.sub.Flush(time.Now().Unix())
```

Comment:
event collection is optional in both the docker and kubernetes integration, would it make sense to also add an option for it in containerd?